Userland/Libraries/LibWeb/WebDriver/ExecuteScript.cpp

```diff
@@ -0,0 +1,363 @@
+        value.has_parameter_map()
+        // - instance of Array
+        || is<JS::Array>(value)
+        // - instance of FileList
+        || is<FileAPI::FileList>(value)
+        // - instance of HTMLAllCollection
+        || false // FIXME
+        // - instance of HTMLCollection
+        || is<DOM::HTMLCollection>(value)
+        // - instance of HTMLFormControlsCollection
+        || false // FIXME
+        // - instance of HTMLOptionsCollection
+        || is<HTML::HTMLOptionsCollection>(value)
+        // - instance of NodeList
+        || is<DOM::NodeList>(value));
+}
+
+// https://w3c.github.io/webdriver/#dfn-json-clone
+static ErrorOr<JsonValue, ExecuteScriptResultType> json_clone(JS::Realm& realm, JS::Value value)
+{
+    // To perform a JSON clone return the result of calling the internal JSON clone algorithm with arguments value and an empty List.
+    auto seen = HashTable<JS::Object*> {};
+    return internal_json_clone_algorithm(realm, value, seen);
+}
+
+// https://w3c.github.io/webdriver/#dfn-internal-json-clone-algorithm
+static ErrorOr<JsonValue, ExecuteScriptResultType> internal_json_clone_algorithm(JS::Realm& realm, JS::Value value, HashTable<JS::Object*>& seen)
+{
+    auto& vm = realm.vm();
+
+    // When required to run the internal JSON clone algorithm with arguments value and seen, a remote end must return the value of the first matching statement, matching on value:
+    // -> undefined
+    // -> null
+    if (value.is_nullish()) {
+        // Success with data null.
+        return JsonValue {};
+    }
+
+    // -> type Boolean
+    // -> type Number
+    // -> type String
+    if (value.is_boolean() || value.is_number() || value.is_string()) {
+        // Success with data value.
+        auto json_string = MUST(JS::JSONObject::stringify_impl(vm, value, JS::js_undefined(), JS::js_undefined()));
+        return MUST(JsonValue::from_string(json_string));
+    }
+
+    // NOTE: BigInt and Symbol not mentioned anywhere in the WebDriver spec, as it references ES5.
+    //       It assumes that all primitives are handled above, and the value is an object for the remaining steps.
+    if (value.is_bigint() || value.is_symbol())
+        return ExecuteScriptResultType::JavaScriptError;
+
+    // FIXME: - a collection
+    // FIXME: - instance of element
+    // FIXME: - instance of shadow root
+    // FIXME: - a WindowProxy object
+
+    // -> has an own property named "toJSON" that is a Function
+    auto to_json = value.as_object().get_without_side_effects(vm.names.toJSON);
+    if (to_json.is_function()) {
+        // Return success with the value returned by Function.[[Call]](toJSON) with value as the this value.
+        auto value_or_error = to_json.as_function().internal_call(value, JS::MarkedVector<JS::Value> { vm.heap() });
+        if (value_or_error.is_throw_completion() || !value_or_error.value().is_string())
+            return ExecuteScriptResultType::JavaScriptError;
```

Comment:
There's a *lot* of manual `is_throw_completion` checking in this, do we want a macro at the top of the file to simplify this?
```c++
#define TRY_JS(expression)                                   \
    ({                                                       \
        auto _temporary_result = (expression);               \
        if (_temporary_result.is_error()) [[unlikely]]       \
            return ExecuteScriptResultType::JavaScriptError; \
        _temporary_result.release_value();                   \
    })
```

Or to return ExecuteScriptResult:
```c++
#define TRY_JS(expression)                                                                          \
    ({                                                                                              \
        auto _temporary_result = (expression);                                                      \
        if (_temporary_result.is_error()) [[unlikely]]                                              \
            return { ExecuteScriptResultType::JavaScriptError, _temporary_result.release_error() }; \
        _temporary_result.release_value();                                                          \
    })
```

Reply:
That may also get rid of some of the `MUST`s, assuming any of them are actually fallible

Reply:
Didn't get rid of all manual checks, but it's an improvement. The `MUST`s were already infallible.